runtime/gc_base/IndexableObjectAllocationModel.cpp

```diff
@@ -321,25 +364,97 @@ MM_IndexableObjectAllocationModel::layoutDiscontiguousArraylet(MM_EnvironmentBas
 	return spine;
 }
 
+/**
+ * Only used if double mapping is enabled
+ */
 #if defined(J9VM_GC_ENABLE_DOUBLE_MAP)
-#if !((defined(LINUX) || defined(OSX)) && defined(J9VM_ENV_DATA64))
-/* Double map is only supported on LINUX 64 bit Systems for now */
-#error "Platform not supported by Double Map API"
-#endif /* !((defined(LINUX) || defined(OSX)) && defined(J9VM_ENV_DATA64)) */
-void * 
-MM_IndexableObjectAllocationModel::doubleMapArraylets(MM_EnvironmentBase *env, J9Object *objectPtr, void *preferredAddress)
+MMINLINE J9IndexableObject *
+MM_IndexableObjectAllocationModel::reserveLeavesForContiguousArraylet(MM_EnvironmentBase *env, J9IndexableObject *spine)
 {
+	Assert_MM_true(_numberOfArraylets == _allocateDescription.getNumArraylets());
+
 	MM_GCExtensions *extensions = MM_GCExtensions::getExtensions(env);
-	J9JavaVM *javaVM = extensions->getJavaVM();
+	GC_ArrayObjectModel *indexableObjectModel = &extensions->indexableObjectModel;
+	const uintptr_t arrayletLeafSize = env->getOmrVM()->_arrayletLeafSize;
+	uintptr_t arrayletLeafCount = MM_Math::roundToCeiling(arrayletLeafSize, _dataSize) / arrayletLeafSize;
+	MM_HeapRegionDescriptorVLHGC *firstLeafRegionDescriptor = NULL;
+#define ARRAYLET_ALLOC_THRESHOLD 64
+	void *leaves[ARRAYLET_ALLOC_THRESHOLD];
+	void **arrayletLeaveAddrs = leaves;
+	if (arrayletLeafCount > ARRAYLET_ALLOC_THRESHOLD) {
+		arrayletLeaveAddrs = (void **)env->getForge()->allocate(arrayletLeafCount * sizeof(uintptr_t), MM_AllocationCategory::GC_HEAP, J9_GET_CALLSITE());
+	}
 
-	GC_ArrayletLeafIterator arrayletLeafIterator(javaVM, (J9IndexableObject *)objectPtr);
-	MM_Heap *heap = extensions->getHeap();
-	UDATA arrayletLeafSize = env->getOmrVM()->_arrayletLeafSize;
-	UDATA arrayletLeafCount = MM_Math::roundToCeiling(arrayletLeafSize, _dataSize) / arrayletLeafSize;
-	Trc_MM_double_map_Entry(env->getLanguageVMThread(), (void *)objectPtr, arrayletLeafSize, arrayletLeafCount);
+	if (NULL == arrayletLeaveAddrs) {
+		return NULL;
+	}
 
-	void *result = NULL;
+	/* determine how many bytes to allocate outside of the spine (in arraylet leaves) */
+	Assert_MM_true(_allocateDescription.getBytesRequested() >= _allocateDescription.getContiguousBytes());
+	uintptr_t bytesRemaining = _allocateDescription.getBytesRequested() - _allocateDescription.getContiguousBytes();
 
+	/* allocate leaf for each arraylet and attach it to its leaf pointer in the spine */
+	uintptr_t arrayoidIndex = 0;
+	while (0 < bytesRemaining) {
+		/* allocate the next arraylet leaf */
+		void *leaf = env->_objectAllocationInterface->allocateArrayletLeaf(env, &_allocateDescription,
+				_allocateDescription.getMemorySpace(), true);
+
```

Comment:
Arraylet leaf is zeroed in the `MM_AllocationContextBalanced::allocateArrayletLeaf()`. This function is called from `MM_TLHAllocationInterface::allocateArrayletLeaf()` if Allocation Context is not NULL.
If I don't miss something we are zeroing leaf unnecessary.